### DISCUSSION TITLE
Project to keep track of typescript improvements

### DISCUSSION THREAD

**Discussion by mattvague** [josdejong/mathjs#2565] [Ideas]
Hey @josdejong @gwhitney, in working on https://github.com/josdejong/mathjs/pull/2537 I had a whole bunch more ideas for how to improve the mathjs typings and am trying to figure out the best way to keep track of them. Wondering if you guys would consider creating a GH project to keep track of that kind of thing that I could add to?

Ideas that I can remember (there were a few more I've forgotten):
- Add a generic to `OperatorNode` to keep track of it's `op` type
- Add generics to `MathArray` and `Matrix` to keep track of their inner value types
- ~~Add dedicated `FractionDefinition` (or something?) type instead of `object` https://github.com/josdejong/mathjs/blob/develop/types/index.d.ts#L543~~
- Stop using union types for `SingularThing | PluralThing`, always use overloads instead
- Convert codebase itself to TS rather than separate typings
- Improve typing consistency, it's a bit all over the place right now (e.g. some things use `BigNumber | number |...` some use `MathNumericType`)
- Figure out how to type return type of `evaluate()`: https://github.com/josdejong/mathjs/blob/develop/types/index.d.ts#L685

**Comment by josdejong** [comment]
> I had a whole bunch more ideas for how to improve the mathjs typings and am trying to figure out the best way to keep track of them

Keep them coming 😄 💪 . 

> Wondering if you guys would consider creating a GH project to keep track of that kind of thing that I could add to?

I'm not sure what you mean? Create a separate project?

I think what will work best is to create a new issue for each idea to improve the TypeScript types, and give them the `typescript` label. That way you can easily search for all open issues/ideas regarding TypeScript. Does that make sense?

**Reply by mattvague** [reply]
Was meaning using the GH projects feature to keep track of the issues.
<img width="1387" alt="image" src="https://user-images.githubusercontent.com/64985/169557081-912f8c5a-b05f-479b-a03b-dec15c92aaf5.png">

> I think what will work best is to create a new issue for each idea to improve the TypeScript types, and give them the typescript label. That way you can easily search for all open issues/ideas regarding TypeScript. Does that make sense?

But yup, that works!

**Reply by josdejong** [reply]
Ah, you mean the new Projects feature of Github. I haven't used that yet. I think it can be really valuable when working with a team on a big project where you need more structure (having deadlines and sprints etc). For mathjs it feels a bit like overkill at this point, let's keep it simple for the time being and just use labeled issues.
